diff --git a/PointDA/data/dataloader.py b/PointDA/data/dataloader.py
index 3e9100c..b4126e1 100644
--- a/PointDA/data/dataloader.py
+++ b/PointDA/data/dataloader.py
@@ -51,9 +51,9 @@ class ScanNet(Dataset):
 
         # split train to train part and validation part
         if partition == "train":
-            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(np.int)
+            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(int)
             np.random.shuffle(self.train_ind)
-            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(np.int)
+            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(int)
             np.random.shuffle(self.val_ind)
 
         io.cprint("number of " + partition + " examples in scannet" + ": " + str(self.data.shape[0]))
@@ -108,9 +108,9 @@ class ModelNet(Dataset):
 
         # split train to train part and validation part
         if partition == "train":
-            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(np.int)
+            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(int)
             np.random.shuffle(self.train_ind)
-            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(np.int)
+            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(int)
             np.random.shuffle(self.val_ind)
 
         io.cprint("number of " + partition + " examples in modelnet : " + str(len(self.pc_list)))
@@ -157,9 +157,9 @@ class ShapeNet(Dataset):
 
         # split train to train part and validation part
         if partition == "train":
-            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(np.int)
+            self.train_ind = np.asarray([i for i in range(self.num_examples) if i % 10 < 8]).astype(int)
             np.random.shuffle(self.train_ind)
-            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(np.int)
+            self.val_ind = np.asarray([i for i in range(self.num_examples) if i % 10 >= 8]).astype(int)
             np.random.shuffle(self.val_ind)
 
         io.cprint("number of " + partition + " examples in shapenet: " + str(len(self.pc_list)))
